client/transactions/list/index.tsx

```diff
@@ -383,20 +388,100 @@ export const TransactionsList = (
 
 	const downloadable = !! rows.length;
 
-	const onDownload = () => {
+	const onDownload = async () => {
+		setIsDownloading( true );
+
 		// We destructure page and path to get the right params.
 		// eslint-disable-next-line @typescript-eslint/no-unused-vars
 		const { page, path, ...params } = getQuery();
+		const downloadType = totalRows > rows.length ? 'endpoint' : 'browser';
+
+		if ( 'endpoint' === downloadType ) {
+			const {
+				date_after: dateAfter,
+				date_before: dateBefore,
+				date_between: dateBetween,
+				match,
+				search,
+				type_is: typeIs,
+				type_is_not: typeIsNot,
+			} = params;
+
+			const isFiltered =
+				!! dateAfter ||
+				!! dateBefore ||
+				!! dateBetween ||
+				!! search ||
+				!! typeIs ||
+				!! typeIsNot;
+
+			const confirmThreshold = 1000;
+			const confirmMessage = sprintf(
+				__(
+					"You are about to export %d transactions. If you'd like to reduce the size of your export, you can use one or more filters. Would you like to continue?",
+					'woocommerce-payments'
+				),
+				totalRows
+			);
 
-		downloadCSVFile(
-			generateCSVFileName( title, params ),
-			generateCSVDataFromTable( columnsToDisplay, rows )
-		);
+			if (
+				isFiltered ||
+				totalRows < confirmThreshold ||
+				window.confirm( confirmMessage )
```

Comment:
Can we add a unit test that covers the else case? Behavioral test for the case when these conditions are not met.

Reply:
Absolutely! Done in https://github.com/Automattic/woocommerce-payments/pull/3477/commits/aa7647f2a055b23da9341ba3413e8d4b059ae13f